File: algorithm/tree/FlattenBinaryTree/FlattenBinaryTree.py
#!/bin/python/
class Solution:
    # @param {TreeNode} root
    # @return {void} Do not return anything, modify root in-place instead.
    def flatten(self, root):
        if root == None:
            return
        
        left  = root.left
        right = root.right
        
        if left is not None:
            root.right = left
            root.left  = None
            rightmost = left
            while rightmost.right != None:
                rightmost = rightmost.right
            rightmost.right = right
        self.flatten(root.right)

File: algorithm/tree/FlattenBinaryTree/FlattenBinaryTree.java
public class FlattenBinaryTree {

  public void flatten(TreeNode root) {
    if (root == null) return;

    TreeNode left   =  root.left;
    TreeNode right  =  root.right;

    if (left != null) {
      root.right = left;
      root.left  = null;
      TreeNode rightmost = left;
      while (rightmost.right != null) {
        rightmost = rightmost.right;
      }
      rightmost.right = right;
    }
    flatten(root.right);
  }

}//end class
